std_execution.bs

```diff
@@ -3032,6 +3174,48 @@ are all well-formed.
 
     3. Returns an operation state `op_state` that contains `op_state2`. When `execution::start(op_state)` is called, calls `execution::start(op_state2)`.
 
+#### `execution::done_as_optional` <b>[execution.senders.adaptors.done_as_optional]</b> #### {#spec-execution.senders.adaptors.done_as_optional}
+
+1. `execution::done_as_optional` is used to handle a done signal by mapping it into the value channel as an empty optional. The value channel is also converted into an optional. The result is a sender that never completes with done, reporting cancellation by completing with an empty optional.
+
+2. The name `execution::done_as_optional` denotes a customization point object. For some subexpression `s.`, let `S` be `decltype((s))`. If the type `S` does not satisfy <code><i>single-typed-sender</i></code>, `execution::done_as_optional(s)` is ill-formed. Otherwise, the expression `execution::done_as_optional(s)` is expression-equivalent to:
+
+    <pre highlight=c++>
+        execution::let_done(
+          execution::then(s, [](auto&& t) {
+              return optional&lt;<i>single-sender-value-type</i>&lt;S>>{
+                static_cast&lt;decltype(t)>(t)
+              };
+            }
+          ),
+          [] {
+            return execution::just(optional&lt;<i>single-sender-value-type</i>&lt;S>>{});
+          }
+        )
+    </pre>
+
+#### `execution::done_as_exception` <b>[execution.senders.adaptors.done_as_exception]</b> #### {#spec-execution.senders.adaptors.done_as_exception}
+
+1. `execution::done_as_exception` is used to handle a done signal by mapping it into the error channel as an `exception_ptr` that refers to a custom exception type. The result is a sender that never completes with done, reporting cancellation by completing with an error.
+
+2. The template <code><i>into-variant-type</i></code> is used to compute the type sent by a sender returned from `execution::into_variant`.
+
+    <pre highlight=c++>
+        template&lt;default_initializable Exception, sender S>
+          <i>see-below</i> done_as_exception(S && s);
+    </pre>
+
+3. <i>Effects:</i> Equivalent to:
+
+    <pre highlight=c++>
+        return execution::let_done(
+          s,
+          [] {
+            return execution::just_error(make_exception_ptr(Exception{}));
```

Comment:
Can we change this so that instead of always constructing a `std::exception_ptr` here, that it just sends `Exception{}` as the error?
The as-awaitable adapter can then turn that into an exception if needed when it throws it.

I'm concerned about the overhead of calling `make_exception_ptr` for cheap error types.

Reply:
Yeah, I think I like that better.